### DISCUSSION TITLE
Tracing Tag Usage

### DISCUSSION THREAD

**Discussion by jgpruitt** [timescale/promscale#873] [Feedback on Features]
OTLP describes attributes which are key value pairs used to decorate various entities in the schema including span, resources, events, and links. In promscale, we chose to refer to these as “tags” rather than attributes as this is a more familiar and easily typed term.

Tags are managed in the database via several tables, types, and operators.

- The `_ps_trace.tag_key` table maintains a distinct list of all the tag keys and provides a surrogate id.
- The `_ps_trace.tag` table contains a row for every tag key value pair ingested. The pairs are given a surrogate id.
- The `ps_trace.tag_map` domain defines a jsonb object in which keys are ids pointing to the `_ps_trace.tag_key` table and values are ids pointing to the `_ps_trace.tag` table. This type should be considered opaque for general usage; operators are defined to make interacting with values of this type easy. The span, link, and event tables have columns of this type to capture the relationship between these entities and the tags they are decorated with.

Operators and functions in the `ps_trace` and `ps_tag` schemas are provided to make interacting with tags easier.

```sql
SELECT
    s.trace_id,
    s.span_id
FROM span s
WHERE s ? ('text7' == 'ee')
;
```

In the above query, the `span` view is aliased as `s`. In the `WHERE` clause, we can refer to `s` and use the `?` operator to filter spans by the tags attached to the spans. `s ?` will consider both span tags and resource tags. The `('text7' == 'ee')` specifies the filtering criterion, and the parentheses are required. `'text7'` specifies the tag key to consider. `==` is a custom operator for use with tags that tests equality. `'ee'` is the the tag value. Thus, this query only returns tags for which there is a tag named "text7" with a value of "ee" either in the span tags or the resource tags associated with the span.

```sql
SELECT
    s.trace_id,
    s.span_id,
    s->'text7' as text7_jsonb
FROM span s
WHERE s ? ('text7' == 'ee')
;
```
```text
               trace_id               |       span_id       | text7_jsonb 
--------------------------------------+---------------------+----------
 2f401ee2-eef0-4e02-958b-a7085f22c28d | 8009739381971939745 | "ee"
 86e785a8-8ec0-438d-8d19-9681fe022bb7 | 4655822916858294253 | "ee"
 5e57e462-c239-4ab3-a48d-8b32505478ef | 4005001813359099333 | "ee"
 8aab3a40-c6a1-4edd-911f-61abd47f44dd | 2629829559601186984 | "ee"
```
Suppose you wish to view the value of the `text7` tag in the results. The `->` operator can be applied to the table alias `s` and the tag key `text7` to return the jsonb value of the tag. Note that the double quotes surrounding the `ee`s are due to the fact the the values are of jsonb type rather than text. To return the value as a text type, use the `->>` operator as below. These operators mimic the [built-in jsonb operators](https://www.postgresql.org/docs/13/functions-json.html#FUNCTIONS-JSON-OP-TABLE).

```sql
SELECT
    s.trace_id,
    s.span_id,
    s->>'text7' as text7_text
FROM span s
WHERE s ? ('text7' == 'ee')
;
```

```text
               trace_id               |       span_id       | text7_text 
--------------------------------------+---------------------+-------
 2f401ee2-eef0-4e02-958b-a7085f22c28d | 8009739381971939745 | ee
 86e785a8-8ec0-438d-8d19-9681fe022bb7 | 4655822916858294253 | ee
 5e57e462-c239-4ab3-a48d-8b32505478ef | 4005001813359099333 | ee
 8aab3a40-c6a1-4edd-911f-61abd47f44dd | 2629829559601186984 | ee
```

We noted that the `?` operator applied to the span view considers both the span tags and resource tags. The same is true of the `->` and `->>` operators; they return the value of the tag if it exists either in the span tags or resource tags (span tags get precedence if the tag exists in both). In many cases, one likely will not care whether a tag is specifically a span tag or resource tag -- just that the tag is associated with the span in some way. However, in the case when the distinction matters, one can be more specific as shown in the queries below. The `val` and `val_text` functions can be used on the individual columns and work like the `->` and `->>` operators respectively. Similarly, the `?` operator can be used to filter against the individual columns too.

```sql
SELECT
    s.trace_id,
    s.span_id,
    val_text(s.span_tags, 'text7') as text7_span,
    val_text(s.resource_tags, 'text7') as text7_resource
FROM span s
WHERE s ? ('text7' == 'ee')
;
```
```text
               trace_id               |       span_id       | text7_span | text7_resource 
--------------------------------------+---------------------+------------+----------------
 2f401ee2-eef0-4e02-958b-a7085f22c28d | 8009739381971939745 |            | ee
 86e785a8-8ec0-438d-8d19-9681fe022bb7 | 4655822916858294253 | ee         | 
 5e57e462-c239-4ab3-a48d-8b32505478ef | 4005001813359099333 | ee         | 
 8aab3a40-c6a1-4edd-911f-61abd47f44dd | 2629829559601186984 | ee         |
```
```sql
SELECT
    s.trace_id,
    s.span_id,
    val_text(s.span_tags, 'text7') as text7_span,
    val_text(s.resource_tags, 'text7') as text7_resource
FROM span s
WHERE s.span_tags ? ('text7' == 'ee')
;
```
```text
               trace_id               |       span_id       | text7_span | text7_resource 
--------------------------------------+---------------------+------------+----------------
 86e785a8-8ec0-438d-8d19-9681fe022bb7 | 4655822916858294253 | ee         | 
 5e57e462-c239-4ab3-a48d-8b32505478ef | 4005001813359099333 | ee         | 
 8aab3a40-c6a1-4edd-911f-61abd47f44dd | 2629829559601186984 | ee         | 
```
```sql
SELECT
    s.trace_id,
    s.span_id,
    val_text(s.span_tags, 'text7') as text7_span,
    val_text(s.resource_tags, 'text7') as text7_resource
FROM span s
WHERE s.resource_tags ? ('text7' == 'ee')
;
```
```text
               trace_id               |       span_id       | text7_span | text7_resource 
--------------------------------------+---------------------+------------+----------------
 2f401ee2-eef0-4e02-958b-a7085f22c28d | 8009739381971939745 |            | ee
```

Tag values support more than just text. Tag values can be booleans, numbers, text, etc. Because they are stored as jsonb values in the database, any value supported by jsonb can be used, however in practice one should limit values to those also supported by the OLTP specification for [Attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/common.md#attributes).

Furthermore, in some cases one does not want to filter by tag values but by the existence of a tag -- i.e. one wants to return any spans that have a `date13` tag associated with them regardless of the value of the `date13` tag. The `?` operator supports this usage pattern as well.

The below query illustrates how one would find all spans with a `date13` tag association and "rendering" the tag value as a `date` datatype.

```sql
SELECT
    s.trace_id,
    s.span_id,
    (s->>'date13')::date as date13
FROM span s
WHERE s ? 'date13'
;
```
```text
               trace_id               |       span_id       |   date13   
--------------------------------------+---------------------+------------
 9eeed6a5-1532-468c-b9b9-d3fb9bef7912 | 1021261185964047190 | 2021-06-14
 f9fdf166-c01f-42b5-8654-b9a93a10c591 | 3551502217513437877 | 2021-05-08
 30e985dd-7b28-4825-8a7f-a6b661baddc9 | 7087634591234281288 | 2021-02-06
 3ac83e2f-48a3-4767-b88e-e838156d3f20 | 7458190695641675701 | 2021-03-02
 879b47ed-500f-43bd-b766-f76fcea3eb02 | 7737180394772223836 | 2021-06-11
 6708e82e-92b9-461f-bcb0-d0cfb23f36de |  307526488779807967 | 2021-02-09
 c7933998-bdc1-4779-a8e7-0c91c0dd8011 | 6002431645986341199 | 2021-09-09
 bae8e998-1748-4c7d-8cb4-7b6de5c47862 | 6639015675319003625 | 2021-08-01
...
 ```
The following query similarly finds spans with an `int42` tag and renders the tag as an `int` type.
```sql
SELECT
    s.trace_id,
    s.span_id,
    (s->'int42')::int as int42
FROM span s
WHERE s ? 'int42'
;
```
```text
               trace_id               |       span_id       | int42 
--------------------------------------+---------------------+-------
 f9fdf166-c01f-42b5-8654-b9a93a10c591 | 4316626515737513386 |    27
 2f401ee2-eef0-4e02-958b-a7085f22c28d | 6329664404467880344 |    50
 2f401ee2-eef0-4e02-958b-a7085f22c28d |  328106973249941925 |    24
 61fae15f-9d4c-42c6-9a48-8181ec73f506 | 3381748025892665184 |     1
 2f401ee2-eef0-4e02-958b-a7085f22c28d |  108324120266379853 |    27
 b1c9929e-2d82-4326-a55c-f45ce7f0b192 | 1588097095266330205 |    44
 b64b3166-5ada-4342-9967-7d0e15c258b1 | 3434751251082907881 |    43
 30e985dd-7b28-4825-8a7f-a6b661baddc9 | 2789624890087032780 |    12
 b1c9929e-2d82-4326-a55c-f45ce7f0b192 | 1443684698392239817 |    31
 b1c9929e-2d82-4326-a55c-f45ce7f0b192 | 6166802474707474229 |    16
 ...
```
To be clear, you may use the capabilities on multiple tag keys at once in the same query. This is fully supported. The below query returns only those spans which have both an `int42` tag AND a `bool5` tag.
```sql
SELECT
    s.trace_id,
    s.span_id,
    (s->'int42')::int as int42,
    (s->'bool5')::bool as bool5
FROM span s
WHERE s ? 'int42'
AND s ? 'bool5'
;
```
```text
               trace_id               |       span_id       | int42 | bool5 
--------------------------------------+---------------------+-------+-------
 b1c9929e-2d82-4326-a55c-f45ce7f0b192 | 6166802474707474229 |    16 | t
 47fe76b2-6096-4ede-ba3d-fd8dcfbc9373 |  990942759936523789 |    30 | f
 6a54237e-2f7e-4f06-88b2-90f70556a761 | 8953670893604374290 |     1 | f
 1e29cc73-402f-4bdf-8167-d51a80e8eec2 | 3088421139965437228 |    37 | f
 7c1b086f-e157-4372-8c72-2a0ee97cf683 | 4691114972447508132 |    20 | f
 9f0a3334-316b-427c-a935-778010d64a3d | 5030628200399810320 |    31 | t
 74115074-694c-4164-ba63-21afd6b91fd8 | 2926175856097105404 |    13 | f
 11eda686-930d-46ba-8270-73d6f2f73180 | 7188821514039705593 |    39 | f
 16680f21-2db4-421f-b9da-361a3a857d0f | 7582840838855527424 |    21 | t
 817b2330-8b20-4255-ab5e-df392ad1aa99 | 3016833593264368524 |    22 | f
 bfce07b2-689b-425b-b542-8b8f16270bf4 | 4648049086883998506 |    10 | f
 1e05a9a0-31ff-4484-9cce-c60c67383531 | 5681672987320304840 |    13 | t
 1e05a9a0-31ff-4484-9cce-c60c67383531 | 8680494626995854159 |    21 | f
 a8e0d3c1-5d24-4829-b7c4-58461cb67cb7 | 7981639470843207938 |    40 | t
 ff7afa11-c1a8-4500-b994-be98ad6cd140 | 7520050808704376784 |    23 | t
```
The `==` operator "understands" many data types that can be used on the right hand side and can apply them correctly to the tag value (which is stored as jsonb). In the below query, spans which have an `int42` tag AND a `bool5` tag with a `false` value are returned.
```sql
SELECT
    s.trace_id,
    s.span_id,
    (s->'int42')::int as int42,
    (s->'bool5')::bool as bool5
FROM span s
WHERE s ? 'int42'
AND s ? ('bool5' == false)
;
```
```text
               trace_id               |       span_id       | int42 | bool5 
--------------------------------------+---------------------+-------+-------
 47fe76b2-6096-4ede-ba3d-fd8dcfbc9373 |  990942759936523789 |    30 | f
 6a54237e-2f7e-4f06-88b2-90f70556a761 | 8953670893604374290 |     1 | f
 1e29cc73-402f-4bdf-8167-d51a80e8eec2 | 3088421139965437228 |    37 | f
 7c1b086f-e157-4372-8c72-2a0ee97cf683 | 4691114972447508132 |    20 | f
 74115074-694c-4164-ba63-21afd6b91fd8 | 2926175856097105404 |    13 | f
 11eda686-930d-46ba-8270-73d6f2f73180 | 7188821514039705593 |    39 | f
 817b2330-8b20-4255-ab5e-df392ad1aa99 | 3016833593264368524 |    22 | f
 bfce07b2-689b-425b-b542-8b8f16270bf4 | 4648049086883998506 |    10 | f
 1e05a9a0-31ff-4484-9cce-c60c67383531 | 8680494626995854159 |    21 | f
```
The following operators are defined for use in filtering tag values:
* `==` equals
* `!==` not equals
* `#<` less than
* `#<=` less than or equal
* `#>` greater than
* `#>=` greater than or equal
* `==~` [regexp matches](https://www.postgresql.org/docs/13/functions-json.html#JSONPATH-REGULAR-EXPRESSIONS)
* `!=~` [regexp not matches](https://www.postgresql.org/docs/13/functions-json.html#JSONPATH-REGULAR-EXPRESSIONS)
* `@?` [jsonb path exists](https://www.postgresql.org/docs/13/functions-json.html#id-1.5.8.22.5.11.2.2.17.1.1.1)

The query below filters integer values using the `#>` and `#<=` operators. It also demonstrates that the filter is considering both the span tags and resource tags.
```sql
SELECT
    s.trace_id,
    s.span_id,
    (s->'int6')::int as int6,
    val(s.span_tags, 'int6')::int as int6_span,
    val(s.resource_tags, 'int6')::int as int6_resource
FROM span s
WHERE s ? ('int6' #> 42)
AND s ? ('int6' #<= 50)
ORDER BY 3
;
```
```text
               trace_id               |       span_id       | int6 | int6_span | int6_resource 
--------------------------------------+---------------------+------+-----------+---------------
 9b5b7537-d5f5-4213-ad14-2ab77903fb51 |  965078591691628920 |    5 |         5 |            46
 1fcbc441-a2ee-46a9-b6fd-3646eb7a1ab0 | 3107325269043663661 |   43 |        43 |              
 f73f8683-4ad5-40fc-9f3e-9fcb78042feb |   68124090128733951 |   43 |           |            43
 ddd94203-0932-44f3-a594-65a96c63c1ce | 1940127402099903838 |   43 |        43 |              
 1496ed07-d40c-4ffa-8abe-8bd667e00b44 |  214377828900615515 |   43 |           |            43
 034e5636-9a56-4c3d-bc9d-9cdd83402a65 | 1606805850479374998 |   43 |        43 |              
 01caa9f6-29b9-4b5b-a4ca-6a22d95d21f1 | 2980290135347879630 |   44 |        44 |              
 07e39d4e-a209-49cd-8d85-5b26e7675b2c | 5717914682119978051 |   45 |           |            45
 ...
 4e4cd353-8923-41e8-b330-d274328a00a2 | 9024028366195800058 |   50 |           |            50
 8f160340-a79e-4441-8ae6-cfb91721661c | 1516171315566012836 |   50 |        50 |            43
 793d2941-4b58-455f-a5bf-a82fea48b8ad | 5791586927086123897 |   50 |        50 |              
 c1c1e2fc-48ff-44dd-9ad0-11d602f9df28 | 4362448605504500042 |   50 |           |            50
 1f581229-eb5d-48b9-9095-99f6e073d3dc | 2240186924701787836 |   50 |        50 |              
```
This query uses the `==~` operator to find tag values that match a regular expression. The query also uses the `#` operator to return the tag id corresponding the value of the `date24` tag.
```sql
SELECT
    s.trace_id,
    s.span_id,
    s # 'date24' as date24_tag_id,
    (s->>'date24')::date as date24
FROM span s
WHERE s ? ('date24' ==~ '2021-03-1.')
ORDER BY 4
;
```
```text
               trace_id               |       span_id       | date24_tag_id |   date24   
--------------------------------------+---------------------+---------------+------------
 739d8e11-12da-437c-b7af-ca6a864c6b94 | 1799904942858251919 |         10444 | 2021-03-10
 c21623e6-1e21-4087-a10e-7afffc152588 | 5389882388679471372 |         12797 | 2021-03-13
 4bec7bc7-69ab-4873-81de-7fc6829b4288 | 2422301284142439872 |         12797 | 2021-03-13
 1aeaef1f-0974-4a71-8e1a-30ab09de0fca | 5011893957897154219 |         28767 | 2021-03-14
 a6b7d343-c981-41fd-ab03-712b4d392080 | 7685197505907553211 |         44865 | 2021-03-15
 54d101c0-4541-4099-b4c9-3a33f7b97330 | 7384761064241742889 |         40184 | 2021-03-16
 a0baaf39-4759-4006-9914-104e61d2efb3 | 5980176931832271346 |         40184 | 2021-03-16
 f234f3a5-ad5d-4caf-96bd-21ff290eaf9d | 5620237367713160301 |          9440 | 2021-03-18
 22f68aa0-261e-4c3f-81de-17a59905a826 | 6872377069778762016 |         25729 | 2021-03-19
```
You can even use the `@?` operator to filter tag values by a [jsonpath query](https://www.postgresql.org/docs/13/functions-json.html#FUNCTIONS-SQLJSON-PATH) to do more powerful evaluations/computations.
```sql
SELECT
    s.trace_id,
    s.span_id,
    (s->>'date24')::date as date24
FROM span s
WHERE s ? ('date24' @? '$ ? (@.datetime() < "2021-01-06".datetime())')
ORDER BY 3
;
```
```text
               trace_id               |       span_id       |   date24   
--------------------------------------+---------------------+------------
 9fc6ac2a-cc9a-4bd8-a704-4cc74bd1a315 |  966648562450143549 | 2021-01-01
 d2f44d5e-7313-4c04-b010-d05e6db03c6e | 7157472989223549186 | 2021-01-01
 2978ed7c-471c-4dd4-9078-b0fe49797ad3 | 3798509444188381201 | 2021-01-01
 7cdde83a-31e2-42cc-bc77-d20f64f2a4cb | 5975310104192654095 | 2021-01-02
 8fd2cfe8-1262-42ec-944d-0f43696ef35f |  609822334500548530 | 2021-01-03
 4e4cd353-8923-41e8-b330-d274328a00a2 | 2486173844502200800 | 2021-01-03
 966239bc-fb89-4a83-97ee-8ef3d812aff3 | 7741438679495867111 | 2021-01-05
 4af7f26f-6927-462c-9650-2897f2d5d32a | 5623698800839195959 | 2021-01-05
```
```sql
SELECT
    s.trace_id,
    s.span_id,
    (s->'int6')::int as int6,
    ((s->'int6')::int + 3) * 2 as is_18
FROM span s
WHERE s ? ('int6' @? '$ ? ((@ + 3) * 2 == 18)')
;
```
```text
               trace_id               |       span_id       | int6 | is_18 
--------------------------------------+---------------------+------+-------
 47fe76b2-6096-4ede-ba3d-fd8dcfbc9373 | 8348106751824798229 |    6 |    18
 3d2066f0-fcf6-4ac4-8511-d0fd861656b3 | 8744861567877718952 |    6 |    18
 5ba53aec-f5b3-46cc-aa36-ada10f3e9fa1 | 2476788606558817390 |    6 |    18
 171500e0-807b-46ff-b435-3c974dc5b560 | 4527742401083684302 |    6 |    18
```

**Comment by ramonguiu** [comment]
Thanks for the super detailed explanation @jgpruitt! This is a great example of how we could write the documentation for this feature.

The solution looks very flexible. Tags will usually be text which means most users will only have to deal with the ? and the -> / ->> operators.

Things get a lot more complicated if you have numerical values or dates in the tags (which will likely not be very common).

A few questions:

1. Using operators on a table is quite strange. Any ideas if we could provide a `tag` column and run the operations against that one?
2. Have you checked how Grafana reacts to jsonb columns in results? What would be a practical example where there is a difference using -> or ->>? In other words, should we just document one of them? For example, would the behavior of order by be what the user expects if using -> ?
3. Do you have to cast the columns? Would there be any way to infer the type from the value when no casting is done?
4. This WHERE clause is extremely complex: `WHERE s ? ('date24' @? '$ ? (@.datetime() < "2021-01-06".datetime())')` . While this will not be a common case (I don't think we should spend a lot of time optimizing for this now) are there any alternatives to be able to do something more consistent with the simpler filtering syntax? Something closer to `WHERE s ? ('int4' < 6)` or `WHERE s ? ('date24' < "2021-01-06".datetime())')`. Would something like `WHERE s ? ('int4'::int < 6)` work?

**Reply by davidkohn88** [reply]
Agreed that this is awesome @jgpruitt! 

I agree on 1, I think it's more intuitive if it's a column rather than the whole row. 
@ramonguiu  3 is a definite you have to cast the columns, Postgres is statically typed and can't do the type of type inference you're asking about unfortunately. 
Agreed on 4, but I think that's just a function of the jsonpath queries and I would very much put that into a separate sort of advanced querying doc that goes over a few of the more advanced query options.

**Reply by jgpruitt** [reply]
>Using operators on a table is quite strange. Any ideas if we could provide a tag column and run the operations against that one?

>I agree on 1, I think it's more intuitive if it's a column rather than the whole row.

Here are the options as I see them:

1. There are two columns -- `span_tags` and `resource_tags` -- and you must interact with them individually
2. There are two columns as above, and the operators work on the view type (what this discussion describes and coded in the `alt-tagging-2` branch)
3. There is one column and the span tags and resource tags are "merged" together in a single map. If the same key is specified at the span and resource level, one of them wins and the other is dropped. There are some issues with operator clashing with the jsonb type.
4. There is one column that contains both the span and resource tags separately. This requires one implementation of the operators for the span table, and a separate implementation for the events and links (which do not have multiple tag_maps). This is coded in the `alt-tagging` branch. There are some issues with operator clashing with the jsonb type.
5. The same as 4, but we coerce the event and link tables to use the same type and operator implementation as the span table. There are some issues with operator clashing with the jsonb type.
6. 🤷🏻‍♂️ I'm am sure there are a dozen more possibilities. @cevian, thoughts?

**Reply by jgpruitt** [reply]
>Have you checked how Grafana reacts to jsonb columns in results? What would be a practical example where there is a difference using -> or ->>? In other words, should we just document one of them? For example, would the behavior of order by be what the user expects if using -> ?

I have not tried it with Grafana yet. `->` returns a jsonb type which may be a **JSON** `string`, `number`, `boolean`, `object`, or `array`. The `->>` operator returns a textual representation of the jsonb value as a **text** type. The distinction is important.

**Reply by jgpruitt** [reply]
>Agreed on 4, but I think that's just a function of the jsonpath queries and I would very much put that into a separate sort of advanced querying doc that goes over a few of the more advanced query options.

Yeah, the jsonpath stuff can be complex, but super powerful. I like the idea of having it explained in a separate "advanced" section.

We might be able to do some more advanced stuff with a simpler syntax, however we'll need to introduce more types to make it happen. For instance, currently there is a `tag_op_greater_than` type that contains the tag_key and filter value for expressions like `('date13' > '2021-01-22'::date)`. At the moment though, the filter value (`2021-01-22`) is of type jsonb. So, the fact that it was originally was a `date` type is lost, and the greater than evaluation is applied alphabetically (as if comparing two strings) rather than chronologically (comparing two dates).

**Reply by thedodd** [reply]
> Yeah, the jsonpath stuff can be complex, but super powerful. I like the idea of having it explained in a separate "advanced" section.

Agreed 100%.

**Comment by cevian** [comment]
Having a tags column that doesn't look super strange on a SELECT * is hard at this point. I want to punt on it.

I just want to point out that 
```
SELECT
    s.trace_id,
    s.span_id,
    (s->'bool5')::bool as bool5
FROM span s
WHERE s ? ('bool5' == false)
```
is equivalent to
```
SELECT
    s.trace_id,
    s.span_id,
    coalesce(s.resource_tags->'bool5', s.span_tags->'bool5')::bool as bool5
FROM span s
WHERE s.resource_tags ? ('bool5' == false) OR s.span_tags ? ('bool5' == false)
```
which in many ways is more "Standard SQL" and is optimized in the same way.

From my point of view the choices we have is either 
1) add the operators on the view s
2) don't add operators and keep things more manual as in the second query above
3) spend more time and see if we could make this work on a column. (this will probably work only with the extension installed).

My preference is between 1 and 2 and NOT 3. probably leaning towards 2 but not strongly.

**Reply by jgpruitt** [reply]
I lean pretty heavily towards option 2 at the moment

**Reply by ramonguiu** [reply]
Why do you prefer 2?

**Reply by jgpruitt** [reply]
I can expound more in the morning, but it's quite a bit less code to maintain, feels more natural, and is more composable/flexible

**Reply by ramonguiu** [reply]
If I focus on the user experience, querying for tags becomes super tedious in my opinion with a lot of repeating code that looks anything but simple.

Maybe heavy users of SQL databases would be ok with it but I doubt the large majority of users will prefer it. We should definitely test it and I am with us adding it later assuming this would just be building on top of what we would have versus requiring data migration.

**Reply by thedodd** [reply]
My instinct is actually to lean towards option `1.`. For me, the overloaded syntax being isolated to a specific view is fine, and I think the simplicity that it brings along with it is quite nice.

I will say though, option `3.` may be a good long-term option, especially with Cloud, as we would be able to more easily prepare customer instances with all of the necessary extensions and such (which we already do).

**Comment by cevian** [comment]
Ok one more option. Along the lines of the view option I had earlier where each tag key becomes it's own column.

What if we had one view per service/operation combo that had all keys as columns. That way we'd get: 
1) some schema discoverability @davidkohn88 mentioned earlier
2) the views would be created automatically.

The only thing is people would need to know which service/operation they are looking at.

We'd need to track which keys occur with which operation but that should be easy and cheap. We'd also track that for event and log entries too but that should be similarly easy. This would allow views for event/logs on specific operations too.

**Reply by cevian** [reply]
so it would look like (for the checkout service, addcart operation):

```
SELECT trace_id, span_id, namespace, pod
FROM span_views.checkout_addcart
WHERE error AND duration_ms >1000 and my_tag='blah'
```

**Reply by cevian** [reply]
1 potential downside:
wouldn't work if users didn't know which service/operation they wanted to analyze (or rather they'd need to fallback to the general span table)

@ramonguiu you didn't like previous view ideas like this because you felt people wouldn't know which one to look at. Do you thin people would know if it was serviice/operation pair?

**Reply by cevian** [reply]
This would have to be in addition to the span table interface described at the top-level since some analysis needs to cross operation boundaries (e.g. which services talk to each other, their rate etc.) On the other hand I'd assume tags are a lot more important for individual operations and cross-operation analysis uses tags less (because not all things will have the same tags)

**Reply by ramonguiu** [reply]
This idea for views makes more sense to me. There are usually two types of queries:

1. General queries to surface things to look at in more detail across services/operations (i.e. one service/operation with abnormal slow response, or with very high error rate)
2. Drill in queries to understand behavior which will usually be on a per-service basis and where the views you suggest would apply.

The challenge is that queries for 1 and 2 would look different and so users would need to learn multiple ways to query the data and they will still need the "non-pure SQL" queries for 1, so they need to learn those anyway.

**Reply by thedodd** [reply]
@cevian I definitely like the example of having a `{service}_{operation}` view, as that simplifies the queries which folks need to execute when, for example, they are hacking on microservice `foo` operation `bar`, and they are interested in evaluating the flow of the operation, ensuring it is working properly, is performant, etc. On the Cloud team, we use that query ALL the time.

I do also think that being able to query for all of the operations per service is important as well (maybe also just for discoverability), as there have been times on the Cloud team where we change the name of an operation, and not everyone will recall exactly the name of the operations. So being able to query first, and then drill into a specific `{service}_{operation}` view would be awesome. And also perfect for building UIs on top of something like this. Of course, psql gives some nice functionality to view DDL, which would also help on this front, though general purpose querying will likely still be important.
